src/vivarium_cluster_tools/psimulate/worker/vivarium_work_horse.py

```diff
@@ -56,68 +148,16 @@ def work_horse(job_parameters: dict) -> tuple[pd.DataFrame, dict[str, pd.DataFra
         if backup:
-            logger.info(f"Restarting simulation from saved backup")
-            sim = backup
-            event["simulation_start"] = time()
-            exec_time = {
-                "setup_minutes": (event["simulation_start"] - event["start"]) / 60
-            }  # execution event
-            logger.info(
-                f'Simulation setup completed in {exec_time["setup_minutes"]:.3f} minutes.'
-            )
+            get_sim_from_backup(event, backup)
         else:
             sim = ParallelSimulationContext(
                 job_parameters.model_specification,
                 configuration=job_parameters.sim_config,
                 logging_verbosity=job_parameters.extras["sim_verbosity"],
             )
-            logger.info("Simulation configuration:")
-            logger.info(str(sim.configuration))
-            sim.setup()
-            event["simulant_initialization_start"] = time()
-            exec_time = {
-                "setup_minutes": (event["simulant_initialization_start"] - event["start"])
-                / 60
-            }  # execution event
-            logger.info(
-                f'Simulation setup completed in {exec_time["setup_minutes"]:.3f} minutes.'
-            )
-
-            sim.initialize_simulants()
-            event["simulation_start"] = time()
-            exec_time["simulant_initialization_minutes"] = (
-                event["simulation_start"] - event["simulant_initialization_start"]
-            ) / 60
-            logger.info(
-                f'Simulant initialization completed in {exec_time["simulant_initialization_minutes"]:.3f} minutes.'
-            )
-        num_steps = sim.get_number_of_steps_remaining()
-        logger.info(f"Starting main simulation loop with {num_steps} time steps")
-        backup_path = (
-            job_parameters.backup_configuration["backup_dir"] / str(get_current_job().id)
-        ).with_suffix(".pkl")
-        sim.run(
-            backup_freq=job_parameters.backup_configuration["backup_freq"],
-            backup_path=backup_path,
-        )
-        event["results_start"] = time()
-        exec_time["main_loop_minutes"] = (
-            event["results_start"] - event["simulation_start"]
-        ) / 60
-        exec_time["step_mean_seconds"] = (
-            event["results_start"] - event["simulation_start"]
-        ) / num_steps
-        logger.info(
-            f'Simulation main loop completed in {exec_time["main_loop_minutes"]:.3f} minutes.'
-        )
-        logger.info(f'Average step length was {exec_time["step_mean_seconds"]:.3f} seconds.')
-
-        sim.finalize()
-        sim.report(print_results=False)
-        event["end"] = time()
-        end_snapshot = CounterSnapshot()
-        do_sim_epilogue(start_snapshot, end_snapshot, event, exec_time, job_parameters)
-        results = sim.get_results()  # Dict[measure, results dataframe]
+            exec_time = initialize_new_sim(event, sim)
```

Comment:
nit: might be more readable if you initialize `sim` within the function as well and then return a tuple `sim, exec_time = initialize_new_sim(event)`

Reply:
If you did that, you could even then change this if/else block to a one-liner: `sim, exec_time = get_sim_from_backup(event, backup) if backup else initialize_new_sim(event, sim)`